clang/lib/CIR/Dialect/Transforms/CIRSimplify.cpp

```diff
@@ -0,0 +1,144 @@
+    return kind == mlir::cir::CastKind::bool_to_int ||
+           kind == mlir::cir::CastKind::int_to_bool ||
+           kind == mlir::cir::CastKind::integral;
+  }
+
+  // makes a chain of casts with the next properties:
+  // - each member is a cast of one of the next kinds: integral, bool_to_int,
+  // int_to_bool
+  // - each member except the last one may have only one use - the next member
+  // in the chain
+  // - last member may have one or many users
+  void collect_casts(chain_of_casts &casts, mlir::cir::CastOp op) const {
+    auto val = op.getResult();
+    if (is_int_or_bool_cast(op)) {
+      casts.push_back(op);
+      if (val.hasOneUse())
+        if (auto next = dyn_cast<mlir::cir::CastOp>(*val.user_begin()))
+          collect_casts(casts, next);
+    }
+  }
+
+public:
+  using mlir::OpRewritePattern<mlir::cir::CastOp>::OpRewritePattern;
+
+  mlir::LogicalResult
+  matchAndRewrite(mlir::cir::CastOp op,
+                  mlir::PatternRewriter &rewriter) const final {
+
+    chain_of_casts casts;
+    collect_casts(casts, op);
+
+    if (casts.size() < 2 || !is_bool_aimed(casts.back()))
+      return mlir::failure();
+
+    auto first = casts.front();
+    auto last = casts.back();
+
+    mlir::Value actual;
+    auto it = casts.begin();
+
+    // The next two cases are covered:
+    // 1) bool_to_int -> ... -> int_to_bool: remove all the casts:
+    //    bool value was already obtained before all the conversions
+    // 2) int_to_bool -> ... -> int_to_bool: take the result of the first op,
+    //    erase all others
+    if (first.getKind() == mlir::cir::CastKind::bool_to_int) {
+      actual = first.getSrc();
+    } else if (first.getKind() == mlir::cir::CastKind::int_to_bool) {
+      actual = first.getResult();
+      it++;
+    } else {
+      return mlir::failure();
+    }
+
+    rewriter.replaceAllUsesWith(last, actual);
+
+    for (auto rit = casts.end() - 1; rit >= it; --rit)
+      rewriter.eraseOp(*rit);
+
+    return mlir::success();
+  }
+};
+
+class SimplifyUnaryNot : public mlir::OpRewritePattern<mlir::cir::UnaryOp> {
+
+  bool is_unary_not(mlir::cir::UnaryOp op) const {
+    return op.getKind() == mlir::cir::UnaryOpKind::Not;
+  }
+
+public:
+  using mlir::OpRewritePattern<mlir::cir::UnaryOp>::OpRewritePattern;
+
+  mlir::LogicalResult
+  matchAndRewrite(mlir::cir::UnaryOp op,
+                  mlir::PatternRewriter &rewriter) const final {
+    if (!is_unary_not(op) || !op.getResult().hasOneUse())
+      return mlir::failure();
+
+    auto val = op.getResult();
+    auto user = *val.user_begin();
+
+    if (auto next = dyn_cast<mlir::cir::UnaryOp>(user)) {
+      if (!is_unary_not(next))
+        return mlir::failure();
+      rewriter.replaceAllUsesWith(next, op.getInput());
+      rewriter.eraseOp(next);
+      rewriter.eraseOp(op);
+      return mlir::success();
+    }
+
+    return mlir::failure();
+  }
+};
+
+void CIRSimplifyPass::runOnOperation() {
+  RewritePatternSet patterns(&getContext());
```

Comment:
If we need to create a new pass (after we go over the other comments), perhaps this is a good opportunity to rename MergeCleanups to CIRSimplify and add this over there instead.

Reply:
done!